### ISSUE TITLE
Create internal Ticket for Customer

### ISSUE THREAD

**Issue by bkolbe** [zammad/zammad#1404]
It would be helpfull if a ticket could be marked as internal, so that we can create tickets for customers without customers to see it. 
We need it to track work done for customers. But we dont want them to receive emails or see them.

Additional it would be good to toggle visitblity of that ticket, to maybe later turn it visible again.

**Comment by voughtdq** [comment]
I've created a rudimentary version of this. The ticket itself is still visible to the customer, but the ticket article is not visible and, like the "Outbound Call" option, it doesn't send an email. All I've done is add an "internal" tab and set the article param for internal to `true`. 

I really just wanted a tab that did what "Outbound Call" does but with a name that makes more sense for my workflow.

Not sure if you would want a pull request. There are no tests and I haven't even added migrations for the additional "internal" option. 

I do want to eventually add the ability to do internal tickets that customer don't see at all, so if you have any advice on where you would start to implement that, I can maybe try to take a crack at it. Does it make sense to add an `internal` attribute to the ticket model in the backend? 

Let me know what you think. 

![Screenshot_20220129_191549](https://user-images.githubusercontent.com/1967287/151682217-6f83894f-8720-4b83-953e-c4a022031f0f.png)
![Screenshot_20220129_191641](https://user-images.githubusercontent.com/1967287/151682221-9614ef5e-71f3-4eaf-8029-48844730b121.png)
![Screenshot_20220129_191735](https://user-images.githubusercontent.com/1967287/151682226-1414cdc3-9dd0-4b1f-9bc5-108205abc1e8.png)

**Comment by erik12321** [comment]
> I've created a rudimentary version of this. The ticket itself is still visible to the customer, but the ticket article is not visible and, like the "Outbound Call" option, it doesn't send an email. All I've done is add an "internal" tab and set the article param for internal to `true`.
> 
> I really just wanted a tab that did what "Outbound Call" does but with a name that makes more sense for my workflow.
> 
> Not sure if you would want a pull request. There are no tests and I haven't even added migrations for the additional "internal" option.
> 
> I do want to eventually add the ability to do internal tickets that customer don't see at all, so if you have any advice on where you would start to implement that, I can maybe try to take a crack at it. Does it make sense to add an `internal` attribute to the ticket model in the backend?
> 
> Let me know what you think.
> 
> ![Screenshot_20220129_191549](https://user-images.githubusercontent.com/1967287/151682217-6f83894f-8720-4b83-953e-c4a022031f0f.png) ![Screenshot_20220129_191641](https://user-images.githubusercontent.com/1967287/151682221-9614ef5e-71f3-4eaf-8029-48844730b121.png) ![Screenshot_20220129_191735](https://user-images.githubusercontent.com/1967287/151682226-1414cdc3-9dd0-4b1f-9bc5-108205abc1e8.png)

Hi voughtdq, 

This is really great! Our company needs this as well. Hope this will be implemented soon and become available in the online versions as well. The simpel version will do nicely as far as I'm concerned. We would use is to schedule maintenance tasks (for instance monthly security updates, backup checks, performance tuning, ..). It is covered in the service contracts and the customer should see it, but not bothered with it. A scheduler is not needed, we would just enter the issues manually. Hope to see this functionality in production soon.

Regards, Erik

**Comment by dominikbenner** [comment]
Hi!

Great function. We also need this in our company.
In addtion to this we must have the option to hide the ticket from the customer completely.

So we need a ticket which is added to customers organization but the customer is not able to see IT

BR

Dominik Benner

**Comment by voughtdq** [comment]
Hi all,

I have some good news, although I'm not sure how the hosted Zammad instances handle this. Zammad has built-in features to allow for internal tickets in the administration interface. It's just not obvious.

The way to implement internal tickets works like this:

1. Modify the Ticket object in the [object manager](https://admin-docs.zammad.org/en/latest/system/objects.html) to have an "Internal" attribute. Click "new attribute", name it "internal" and select the boolean option.
2. If you have any triggers that send new tickets/updates to customers, you will want to modify those triggers to only work when "internal" is "no" (false).
3. For any customer facing overviews, you will want to modify those overviews to only include tickets whose "internal" attribute is "no" (false).

This effectively creates internal tickets for your Zammad instance. The screenshots I posted are mostly cosmetic. One thing to watch out for would be the default behavior for articles on new tickets. The articles are public, but my understanding is that, because the customer overviews will filter out tickets where internal is true, they shouldn't be able to see the tickets.

**Comment by erik12321** [comment]
Hi Douglas,

Thanks for your workaround!

The solution remains confusing, I think. It is no “Receive Call”, it is no “Outbound call”  and no “Send Email”.

It would be nice if we had an additional button like “internal” or other as depicted below.
***@***.***

In our situation more than 90% of our tickets would start as an “internal” ticket!

What do you think?

Regards,
Erik

**Comment by itkfm** [comment]
> This effectively creates internal tickets for your Zammad instance.

nope, those will be hidden, not really internal, unfortunately

> they shouldn't be able to see the tickets.

they can still *access* them, assuming they guess the ID of the hidden tickets

**Comment by MrGeneration** [comment]
Please move workaround or configuration topics to the Zammad community to reduce the noise and this issue tracker. (https://community.zammad.org)

Workaround discussions unfortunately do not improve this issues situations nor speeds it up.
Thank you!

**Comment by voughtdq** [comment]
> they can still _access_ them, assuming they guess the ID of the hidden tickets

Good point. I'll take a closer look at the permission system to get a better idea about how to implement access control to make them truly internal. I wasn't too worried about this because it wouldn't be an issue for *my* customers to see the tickets by guessing the IDs. I understand that for others, this is not the case.

> Workaround discussions unfortunately do not improve this issues situations nor speeds it up.

I disagree. I just got vital information about lack of consideration for ticket permissions. I've also gotten feedback and encouragement from two people who want this feature. I'm not confident I would have received any feedback had I posted this on the forum. There doesn't seem to be much engagement there; the pinned thread on the Development forum is just placeholder text. This is also no longer a workaround, but something that will require modification to the codebase, for which I will have to develop tests.

I'll add any additional progress as a pull request.

**Comment by chrisvanberg** [comment]
Hello, 
I'm not sure if this message should be here or on a new thread on the community forum.

However I was wondering if you have any tips on how to replicate the "Internal button" while using the solution with the objet "Internal" on tickets ? 
![](https://user-images.githubusercontent.com/1967287/151682217-6f83894f-8720-4b83-953e-c4a022031f0f.png)

We run the self-hosted version and have already implemented the internal field on tickets, but the internal button on ticket creation could really be usefull for our agents.

We are also open to work on a PR, but we would need some guidances 🙂

**Comment by voughtdq** [comment]
The frontend surprisingly needs only a few changes to do the internal button.

https://github.com/zammad/zammad/blob/6.6.0-alpha/app/assets/javascripts/app/controllers/agent_ticket_create.coffee#L30-L46

And then a tweak here to set `internal: true` for the article params. 

https://github.com/zammad/zammad/blob/6.6.0-alpha/app/assets/javascripts/app/controllers/agent_ticket_create.coffee#L658-L705

I'd need to go back and look at the specific tweaks I made because I found that internal tickets aren't really something I use in practice, so the code is in some archive somewhere.

The bigger problem is making the appropriate changes to the permissions system and adding tests. I'm unfamiliar with how they work.

**Comment by chrisvanberg** [comment]
Thanks for your response and links!

#### For reference
It seems that what we were looking for was a way to create a ticket starting **with an internal note**.

That way, we are sure that **the customer will never see the note and won't be notified with its content** since it's an internal note. Even if the customer guesses the ticket id and access the ticket with the UI 🎉

The visibility (in the overview pane) of the ticket itself for the customer is still managed with @voughtdq approach.

#### Based on our analysis:
At the moment, **it seems only possible with an API call** and sadly not in the UI, so when we need to create "internal note" ticket, we use this call on `{{zammad_base_url}}/tickets/`

With this body :
```json
{
  "title": "{{ticket_title}}",
  "group": "{{ticket_group}}",
  "customer_id": "guess:{{customer_email}}",
  "owner_id": "{{zammad_agent_id}}",
  "article": {
    "body": "{{ticket_body}}",
    "type": "note",
    "internal": true
  },
  "state_id": 2,
}
```

#### What we could try
> The frontend surprisingly needs only a few changes to do the internal button. ([link](https://github.com/zammad/zammad/blob/6.6.0-alpha/app/assets/javascripts/app/controllers/agent_ticket_create.coffee#L30-L46))
> And then a tweak here to set internal: true for the article params.

It seems that this solution (creating a ticket starting with an internal note) does not need to modify the "backend" of Zammad, only the "frontend".

I'll try that on our instance and then see if we could make a PR.

**Comment by sjacobflaherty** [comment]
A zammad solution for this would be great. In our usecase, we might want to email _somewhere_ down the line, and keeping all fields and information we have attached to the same ticket would be amazing.

**Comment by dansik8** [comment]
This would be an amazing feature. I would really appriciate it.
